Feat: Add Langtrace integration

alizenhom opened the pull request:
## Title
Add Langtrace Integration for Litellm

<!-- e.g. "Implement user authentication feature" -->

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🆕 New Feature


## Changes

<!-- List of changes -->

## [REQUIRED] Testing - Attach a screenshot of any new tests passing locall
![Screenshot 2024-08-23 at 10 11 46 PM](https://github.com/user-attachments/assets/cf2adc1a-7f0f-46ed-a927-c5668a2b7278)If UI changes, send a screenshot/GIF of working UI fixes

<!-- Test procedure -->

alizenhom committed:
Feat: Add Langtrace integration

alizenhom committed:
add langtrace service name

alizenhom committed:
fix timestamps for traces

alizenhom committed:
add tests

krrishdholakia commented on litellm/litellm_core_utils/litellm_logging.py:
```diff
@@ -1090,6 +1091,27 @@ def success_handler(
                             end_time=end_time,
                             print_verbose=print_verbose,
                         )
+                    if callback == "langtrace":
```
hey @alizenhom i don't think we need an explicit callback, since you're implementation is CustomLogger compatible 

this should be handled here - https://github.com/BerriAI/litellm/blob/3007f0344d3a84c019377a27935bcc786bdaf253/litellm/litellm_core_utils/litellm_logging.py#L1947

alizenhom replied:
agreed, will remove it from callbacks and handle it as mentioned

krrishdholakia commented on litellm/integrations/langtrace.py:
```diff
@@ -0,0 +1,143 @@
+    def __init__(self) -> None:
+        try:
+            from langtrace_python_sdk import langtrace
+            from opentelemetry.trace import get_tracer
```
@alizenhom we already have an OTEL Logger - https://github.com/BerriAI/litellm/blob/3007f0344d3a84c019377a27935bcc786bdaf253/litellm/integrations/opentelemetry.py#L41

If langtrace is otel compatible, the only thing we need is to show how to use it in docs - https://docs.litellm.ai/docs/proxy/logging#logging-proxy-inputoutput-in-opentelemetry-format

alizenhom replied:
I see, i will clean all this up 
1. use existing Otel Logger
2. will keep the extra file just to add extra attributes needed by langtrace
3. will adjust otel logger, if `callback_name` is `langtrace` then call `langtrace_set_attributes`

krrishdholakia reviewed (changes requested):
If langtrace is otel compatible, a separate file might not be required. 

We just need to show how to use it in docs - https://docs.litellm.ai/docs/proxy/logging#logging-proxy-inputoutput-in-opentelemetry-format

alizenhom committed:
Discard Callback + use existing otel logger

alizenhom committed:
cleanup

alizenhom committed:
remove print statments

alizenhom committed:
remove callback

alizenhom committed:
add docs

alizenhom committed:
Merge branch 'BerriAI:main' into add-langtrace-integration

alizenhom committed:
docs

alizenhom committed:
add logging docs

alizenhom commented:
@krrishdholakia changes resolved kindly review once again

alizenhom committed:
Merge remote-tracking branch 'upstream/main' into add-langtrace-integration

alizenhom committed:
format logging

ishaan-jaff commented on docs/my-website/docs/observability/langtrace_integration.md:
```diff
@@ -0,0 +1,43 @@
+import Image from '@theme/IdealImage';
+
+# 🔥 Langtrace AI
```
no emojis please

ishaan-jaff commented on docs/my-website/docs/observability/langtrace_integration.md:
```diff
@@ -0,0 +1,43 @@
+
+## Pre-Requisites
+Make an account on [Langtrace AI](https://langtrace.ai/login)
+
```
Add a link / example of how to use with LiteLLM proxy - majority of our users use the proxy @alizenhom

ishaan-jaff commented on docs/my-website/docs/proxy/logging.md:
```diff
@@ -61,9 +61,9 @@ litellm_settings:
 
 Removes any field with `user_api_key_*` from metadata.
 
-## What gets logged? StandardLoggingPayload
```
there are many changes made to this doc - can you please scope the doc changes just to adding langtrace

ishaan-jaff commented on docs/my-website/docs/proxy/logging.md:
```diff
@@ -61,9 +61,9 @@ litellm_settings:
+## What gets logged?
 
-Found under `kwargs["standard_logging_object"]`. This is a standard payload, logged for every response.
+Found under `kwargs["standard_logging_payload"]`. This is a standard payload, logged for every response.
```
this is wrong it is actually `kwargs["standard_logging_object"]`

ishaan-jaff reviewed (changes requested):
minor changes - code looks good @alizenhom

alizenhom commented:
thanks for the feedback, will address these changes now

alizenhom committed:
remove emoji and add litellm proxy example

alizenhom committed:
format logging

alizenhom committed:
format `logging.md`

alizenhom committed:
add langtrace docs to logging.md

alizenhom committed:
sync conflict

krrishdholakia commented on docs/my-website/docs/observability/langtrace_integration.md:
````diff
@@ -0,0 +1,63 @@
+```python
+import litellm
+import os
+from langtrace_python_sdk import langtrace
````
why is `langtrace` required if it just uses our otel integration? @alizenhom

krrishdholakia commented on litellm/tests/test_langtrace.py:
```diff
@@ -0,0 +1,33 @@
+
+@pytest.mark.parametrize("model", ["claude-2.1", "gpt-3.5-turbo"])
+def test_langtrace_logging(exporter, model):
+    litellm.completion(
```
i just ran this test. Nothing gets logged to langtrace.
<img width="1695" alt="Screenshot 2024-10-12 at 4 12 16 PM" src="https://github.com/user-attachments/assets/9f83cb80-1af2-4e94-a744-f61200a21d23">

krrishdholakia commented:
I just ran the test. 

I can confirm our otel integration is called. 

Nothing is shown on langtrace.

krrishdholakia commented:
I suspect their client sdk is monkeypatching otel 

When i comment it out, i get the error from otel 

> Failed to export batch code: 404, reason: {"error":"Something went wrong while ingesting traces"}

krrishdholakia commented:
@alizenhom can you please submit a fix with working langtrace logging

if your logging destination is otel compatible, then please submit a fix without the langtrace sdk. 

This will make it easier for proxy users to try langtrace without needing to create custom dockerfiles.

alizenhom commented:
hey @krrishdholakia  
kindly check #6259 out it addresses all comments, 

also we found a small issue from our side that we fixed and releasing today, i have tested both sdk and proxy examples and works like a charm with langtrace.

once we merge our fix today, litellm users will be able to ship to langtrace seamlessly both using otel or langtrace callbacks
